resources/aws/auto_scaling_group.go

```diff
@@ -161,3 +164,99 @@ func (asg *AutoScalingGroup) findExisting() (*autoscaling.Group, error) {
 
 	return autoScalingGroups.AutoScalingGroups[0], nil
 }
+
+// FindLegacy returns true if there is an ASG created for the same cluster name
+// but not being part of a CloudFormation stack.
+func (asg *AutoScalingGroup) FindLegacy() (bool, error) {
+	a, err := asg.findExisting()
+	if IsNotFound(err) {
+		return false, nil
+	} else if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	return isLegacyASG(asg.Name, asg.ClusterID, a.Tags), nil
+}
+
+// DetachInstances detaches instances from this ASG and returns their details
+func (asg *AutoScalingGroup) DetachInstances() ([]*autoscaling.Instance, error) {
+	autoScalingGroup, err := asg.findExisting()
+	if err != nil {
+		return []*autoscaling.Instance{}, microerror.Mask(err)
+	}
+
+	for _, i := range autoScalingGroup.Instances {
+		operation := func() error {
+			_, err = asg.Client.DetachInstances(&autoscaling.DetachInstancesInput{
+				AutoScalingGroupName:           aws.String(asg.Name),
+				InstanceIds:                    []*string{i.InstanceId},
+				ShouldDecrementDesiredCapacity: aws.Bool(false),
+			})
+			return err
+		}
+		err := backoff.Retry(operation, backoff.NewExponentialBackOff())
```

Comment:
We usually use `RetryNotify` and log retry error. Otherwise on retries it looks like the pod is hanging. In order to achieve that you'll have to pass logger to AutoSaclingGroup.

Reply:
Example: https://github.com/giantswarm/operatorkit/blob/14cddd8e1d187249ff6a47307036a3c14f7052d7/framework/retryresource/resource.go#L94-L101

Reply:
Thanks, done. I'll retry the full upgrade test on local example now that all the changes are in place